### PR TITLE
CA-294634: Fix issue of finding multipath config

### PR DESCRIPTION
**Pull request by DeliZhangX**
Signed-off-by: Deli Zhang <deli.zhang@citrix.com>

### CODE REVIEW

**Comment by DeliZhangX** [comment]
XenRT jobs passed:
* iSCSI: 2747887
* HBA: 2747889

File: src/XenCert/StorageHandlerUtil.py [hunk 1]
```diff
@@ -661,59 +661,90 @@ def get_lun_scsiid_devicename_mapping(targetIQN, portal):
...
-                regexvendor = re.compile(vendorSearch)
-                regexproduct = re.compile(productSearch)
-                if ((regexproduct.search(product)) and (regexvendor.search(vendor))):
-                    break
-            else:
-                skipThis = False
+        d = parse_multipathd_config([line+'\n' for line in stdout.split('\n')])
+        XenCertPrint("mpath config to dict: %s" % d)
+
+        for _,device_value in d["devices"]:
```
Comment (minli1) [inline src/XenCert/StorageHandlerUtil.py]: should have indent after the comma

File: src/XenCert/StorageHandlerUtil.py [hunk 2]
```diff
@@ -661,59 +661,90 @@ def get_lun_scsiid_devicename_mapping(targetIQN, portal):
...
+            attr_map = dict(device_value)
+            if 'vendor' not in attr_map or 'product' not in attr_map:
+                XenCertPrint("warning: skip the device attributes because can not find mandatory key vendor or product")
+                continue
+            re_vendor = re.compile(attr_map['vendor'].strip('"'))
+            re_product = re.compile(attr_map['product'].strip('"'))
+            if (re_vendor.search(vendor) and re_product.search(product)):
+                XenCertPrint("matched vendor and product")
+                device_config = dict(multiPathDefaultsMap.items() + attr_map.items())
+                break
```
Comment (minli1) [inline src/XenCert/StorageHandlerUtil.py]: actually the `break` is not necessary here.

File: src/XenCert/StorageHandlerUtil.py [hunk 3]
```diff
@@ -661,59 +661,90 @@ def get_lun_scsiid_devicename_mapping(targetIQN, portal):
         XenCertPrint("Failed to find any LUNs for IQN: %s and portal: %s" % (targetIQN, portal))
         return {}
 
+def parse_multipathd_config(lines):
```
Comment (minglumlu) [inline src/XenCert/StorageHandlerUtil.py]: The definition of "multipath.conf" is recursively. Would this make the the parsing simpler?
For example:
```
def getNext(confStr):
    if conf_str.startswith("{"):
        word = getFirstWord(conStr)
        return word, confStr[len(word):]
    else:
        idx = getMatched(confStr)
        return confStr[0:idx+1], confStr[(idx+1):]
        

def parse_multipathd_config(confStr, outputDict):
    key, confStr = getNext(confStr)
    value, confStr = getNext(confStr)
    if isValue(value):
        outputDict[key] = value
    else:
        r = parse_multipathd_config(value,{})
        output_dict[key] = r

    return outputDict
```

Comment (DeliZhangX) [reply src/XenCert/StorageHandlerUtil.py]: It looks shorter indeed, while I don't think it's simpler. I am glad to change to use your way if it has better maintainability. let's touch tomorrow.

File: src/XenCert/StorageHandlerUtil.py [hunk 4]
```diff
@@ -661,59 +661,90 @@ def get_lun_scsiid_devicename_mapping(targetIQN, portal):
...
+        for _,device_value in d["devices"]:
+            XenCertPrint("device attributes: %s" % device_value)
+            attr_map = dict(device_value)
+            if 'vendor' not in attr_map or 'product' not in attr_map:
+                XenCertPrint("warning: skip the device attributes because can not find mandatory key vendor or product")
+                continue
+            re_vendor = re.compile(attr_map['vendor'].strip('"'))
+            re_product = re.compile(attr_map['product'].strip('"'))
+            if (re_vendor.search(vendor) and re_product.search(product)):
+                XenCertPrint("matched vendor and product")
```
Comment (XihuanYang) [inline src/XenCert/StorageHandlerUtil.py]: Why all sentence in XenCertPrint all start with lower case?

**Comment by minglumlu** [comment]
LGTM.

File: src/XenCert/StorageHandlerUtil.py (hunk 3 above)
Comment (minglumlu) [reply src/XenCert/StorageHandlerUtil.py]: That's fine.

**Comment by DeliZhangX** [comment]
Already confirmed all items face by face!
Thanks for all your reviewing.

**Pull request merged**